Introduce themeswitcher

pascalwengerter opened the pull request:
## Description
This is currently WIP, but allows users to switch between a dark and light theme (provided this is found in the theming configuration). It initializes with the browser preference

## Related Issue
- Fixes #6242

## Types of changes
- [X] New feature (non-breaking change which adds functionality)
- [X] Technical debt (removes workarounds related to theming/colors in the redesign)

## Checklist:
- [X] Code changes
- [ ] Update existing tests
- [ ] Add new tests
- [X] Add changelog item
- [ ] Save current theme to localStorage
- [X] Change icon/themeSwitcher component location
- [ ] Add docs related to dark theme

tbsbdr commented:
👏
I propose the following switch:
switch shows the currently applied theme (not the to-be-switched theme); if your are in the dark mode, switch label is "Dark mode 🌚"
![2022-01-13_16-04-45 (1)](https://user-images.githubusercontent.com/26610733/149355394-b90128f9-760a-4071-b6ae-be554c8b922e.gif)

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/owncloud/web?pullRequest=6240) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/owncloud/web?pullRequest=6240) before we can accept your contribution.<br/>**1** out of **2** committers have signed the CLA.<br/><br/>:white_check_mark: pascalwengerter<br/>:x: tbsbdr<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/owncloud/web?pullRequest=6240) it.</sub>

kulmann commented on packages/web-runtime/src/helpers/theme.js:
```diff
@@ -17,6 +17,7 @@ export const loadTheme = async (location = '') => {
       return defaults
     }
     const theme = await response.json()
+    console.log('theme', theme)
```
dev leftover (well... still WIP... but shouldn't be committed)

kulmann commented on packages/web-runtime/src/store/config.js:
```diff
@@ -109,7 +118,7 @@ const getters = {
     return parsed
   },
   theme: (state) => {
```
Is the theme getter used at all? You were accessing the theme with `configuration.currentTheme` everywhere. If that's intended you might want to drop this getter entirely. Or map & use it instead of `configuration`.

pascalwengerter commented:
Could someone thoroughly test this with a locally linked https://github.com/owncloud/owncloud-design-system/pull/1886 ? Rebased, animations fixed, extracted some things and fixed stuff the theme switch revealed in the ODS (so we'll need another release there)

kulmann reviewed (changes requested):
Things that I found in the UI while testing:

- (~~search bar is gone?~~ sorry for the noise, was in ocis on a page that has no filtering)
- can we have a small right margin for the left sidebar, so that (when the app has a left sidebar nav) the app content gets a little more distance from the sidebar? At the moment it looks too compact IMO.
- when switching the theme the left sidebar has a slight delay / takes a little bit longer, compared to the rest of the UI

Dark mode:
- form input fields background color is a dark blue, I'd have expected something gray-ish (was looking at e.g. invite input field for people sharing and the page size dropdown in view options)
- OcDrop background same or too similar to the main background color, drop shadow exists but has no to little effect
- "+ New" button has black text on blue button but should have white text afaik
- the left sidebar nav items don't have the thick shadow on the right side of the nav icon
- OcSwitch background in enabled state has very little contrast to the main background color
- mediaviewer shows the label `1 of 3` in black instead of white
- modals look broken

Light mode:
found nothing ✅

pascalwengerter commented:
> Things that I found in the UI while testing:
> 
>     * (~search bar is gone?~ sorry for the noise, was in ocis on a page that has no filtering)
> 
>     * can we have a small right margin for the left sidebar, so that (when the app has a left sidebar nav) the app content gets a little more distance from the sidebar? At the moment it looks too compact IMO.
> 
>     * when switching the theme the left sidebar has a slight delay / takes a little bit longer, compared to the rest of the UI
> 
> 
> Dark mode:
> 
>     * form input fields background color is a dark blue, I'd have expected something gray-ish (was looking at e.g. invite input field for people sharing and the page size dropdown in view options)
> 
>     * OcDrop background same or too similar to the main background color, drop shadow exists but has no to little effect
> 
>     * "+ New" button has black text on blue button but should have white text afaik
> 
>     * the left sidebar nav items don't have the thick shadow on the right side of the nav icon
> 
>     * OcSwitch background in enabled state has very little contrast to the main background color
> 
>     * mediaviewer shows the label `1 of 3` in black instead of white
> 
>     * modals look broken
> 
> 
> Light mode: found nothing white_check_mark
> 
> Edit:
> 
>     * OcDrop styling is part of another story, if there are no easy fixes here, then keep it for later.

Thanks for testing @kulmann, did you try it with https://github.com/owncloud/owncloud-design-system/issues/1887 locally linked?

kulmann commented:
> Thanks for testing @kulmann, did you try it with [owncloud/owncloud-design-system#1887](https://github.com/owncloud/owncloud-design-system/issues/1887) locally linked?

Yes 😇

pascalwengerter commented:
> Things that I found in the UI while testing:
> 
>     * (~search bar is gone?~ sorry for the noise, was in ocis on a page that has no filtering) 
> 
✅ 

>     * can we have a small right margin for the left sidebar, so that (when the app has a left sidebar nav) the app content gets a little more distance from the sidebar? At the moment it looks too compact IMO.
> 
✅ rather added a padding to the router-view, right margin on the sidebar messed with the border-shadow

>     * when switching the theme the left sidebar has a slight delay / takes a little bit longer, compared to the rest of the UI
> 
✅ added the same to the app content area now, please check if it's to your liking

> 
> Dark mode:
> 
>     * form input fields background color is a dark blue, I'd have expected something gray-ish (was looking at e.g. invite input field for people sharing and the page size dropdown in view options)
> 
✅ now has one of the "standard dark theme background colors", similar to the form field background in light mode

>     * OcDrop background same or too similar to the main background color, drop shadow exists but has no to little effect
> 
✅ not to resolve in this ticket, can't find the exalated ticket yet though

>     * "+ New" button has black text on blue button but should have white text afaik
> 
✅ should be resolved now

>     * the left sidebar nav items don't have the thick shadow on the right side of the nav icon
> 
✅ resolved

>     * OcSwitch background in enabled state has very little contrast to the main background color
> 
https://github.com/owncloud/owncloud-design-system/issues/1887, ✅ resolved by changing colors & adding a small border around it

>     * mediaviewer shows the label `1 of 3` in black instead of white
> 
✅ resolved

>     * modals look broken
✅ resolved in the ODS, needed a background color

> 
> 
> Light mode: found nothing white_check_mark
> 
> Edit:
> 
>     * OcDrop styling is part of another story, if there are no easy fixes here, then keep it for later.

Commits for review incoming

pascalwengerter commented:
@kulmann what do you think in terms of a11y, add a tooltip&aria-label to the button and make the text aria hidden?

kulmann reviewed (changes requested):
Some more minor things:

## dark mode
- pagination has dark colors / no contrast
- notification bell has dark colors / no contrast
- search indicators in table header has dark colors / no contrast
- people invite form field has dark text color in autocomplete items (usernames and sharee type)
- OcSwitch color contrast is still bad

pascalwengerter commented:
> Some more minor things:
> ## dark mode
> 
>     * pagination has dark colors / no contrast
> 
✅ consider it fixed in ODS
>     * notification bell has dark colors / no contrast
> 
✅ consider it fixed in web
>     * search indicators in table header has dark colors / no contrast
> 
You mean sorting buttons? That's a tough one since we render (colored) SVGs here...

>     * people invite form field has dark text color in autocomplete items (usernames and sharee type)
> 
I don't find it too bad, what would you like to change?
![Screenshot from 2022-01-18 15-45-28](https://user-images.githubusercontent.com/16822008/149965931-6de415ef-5501-4d31-a286-9b4962aa61d8.png)



>     * OcSwitch color contrast is still bad
Any suggestions how to go about it?

kulmann commented:
> > ```
> > * search indicators in table header has dark colors / no contrast
> > ```
> 
> You mean sorting buttons? That's a tough one since we render (colored) SVGs here...

Yes, the sorting buttons. Would suggest to add a white outline to the icons, so that they are compatible with light and dark mode at the same time.


> > ```
> > * people invite form field has dark text color in autocomplete items (usernames and sharee type)
> > ```
> 
> I don't find it too bad, what would you like to change? ![Screenshot from 2022-01-18 15-45-28](https://user-images.githubusercontent.com/16822008/149965931-6de415ef-5501-4d31-a286-9b4962aa61d8.png)

True, keep it like it is.


> > ```
> > * OcSwitch color contrast is still bad
> > ```
> 
> Any suggestions how to go about it?

Same as for the sorting icons, add a white outline, so that it's compatible with light and dark mode at the same time. The knob in the middle is not that easily visible as well, though.

kulmann reviewed (approved):
Unit tests need fixing, but otherwise this is gold 🥇 happy to get it merged asap!

fschade reviewed (approved):
❤️  LGTM

ownclouders commented:
Results for <strong>oCISTrashbinUploadMoveJourney</strong> https://drone.owncloud.com/owncloud/web/22016/66/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.

ownclouders commented:
Results for <strong>oC10iPhone1</strong> https://drone.owncloud.com/owncloud/web/22017/47/1
:boom: The acceptance tests pipeline failed. The build has been cancelled.
